grove/tomography/utils.py

```
@@ -0,0 +1,640 @@
+    dist = np.cumsum(probs)
+    rs = np.random.rand(n)
+    return np.array([(np.where(r < dist)[0][0]) for r in rs])
+
+
+def basis_state_preps(*qubits):
+    """
+    Generate a sequence of programs that prepares the measurement
+    basis states of some set of qubits in the order such that the qubit
+    with highest index is iterated over the most quickly:
+    E.g., for qubits=(0, 1), it returns the circuits::
+
+        I_0 I_1
+        I_0 X_1
+        X_0 I_1
+        X_0 X_1
+    :param list qubits: Each qubit to include in the basis state preparation.
+    :return: Yields programs for each basis state preparation.
+    :rtype: Program
+    """
+    for prep in cartesian_product([I, X], repeat=len(qubits)):
+        basis_prep = Program()
+        for gate, qubit in zip(prep, qubits):
+            basis_prep.inst(gate(qubit))
+        yield basis_prep
+
+
+def basis_labels(n):
+    """
+    Generate a list of basis labels for `n` qubits, ordered from least to greatest, in big-endian
+     format:
+
+        ['00..00', '00..01', ..., '11..11']
+
+    :param n:
+    :return: A list of strings of length n that enumerate the n-qubit bitstrings
+    :rtype: list
+    """
+    return ["".join(labels) for labels in itertools.product('01', repeat=n)]
+
+
+def sample_bad_readout(program, num_samples, assignment_probs, cxn):
+    """
+    Generate `n` samples of measuring all outcomes of a Quil `program`
+    assuming the assignment probabilities `assignment_probs` by simulating the
+    wave function on a reference-qvm SyncConnection `cxn`
+
+    :param pyquil.quil.Program program: The program.
+    :param int num_samples: The number of samples
+    :param numpy.ndarray assignment_probs: A matrix of assignment probabilities
+    :param reference_qvm.api.SyncConnection cxn: the QVM connection.
+    :return: The resulting sampled outcomes from assignment_probs applied to cxn, one dimensional.
+    :rtype: numpy.ndarray
+    """
+    wf = cxn.wavefunction(program)
```

in the tomography we allow QPU connections, we should guard this here.

---

What do you mean? Do you want to throw an error? Isn't it sufficient for an error to be raised saying that QPU connections don't have a wavefunction method?